packages/unchained-client/src/utxo/parser/thorchain.ts

```diff
@@ -0,0 +1,249 @@
+import type { Axios } from 'axios'
+import axios from 'axios'
+
+import { type BaseTxMetadata, Dex, TradeType } from '../../types'
+import type { SubParser, Tx, TxSpecific } from '../parser'
+
+export type LiquidityType = 'Savers' | 'LP'
+export type SwapType = 'Standard' | 'Streaming'
+
+interface Liquidity {
+  type: LiquidityType
+}
+
+interface Swap {
+  type: SwapType
+}
+
+export interface TxMetadata extends BaseTxMetadata {
+  parser: 'thorchain'
+  memo: string
+  liquidity?: Liquidity
+  swap?: Swap
+}
+
+interface Coin {
+  amount: string
+  asset: string
+}
+
+interface InOut {
+  txID: string
+  address: string
+  coins: Coin[]
+  height?: string
+}
+
+interface Metadata {
+  addLiquidity?: {
+    liquidityUnits: string
+  }
+  refund?: {
+    affiliateAddress: string
+    affiliateFee: string
+    memo: string
+    networkFees: Coin[]
+    reason: string
+  }
+  swap?: {
+    affiliateAddress: string
+    affiliateFee: string
+    isStreamingSwap: boolean
+    liquidityFee: string
+    networkFees: Coin[]
+    swapSlip: string
+    swapTarget: string
+    memo: string
+  }
+  withdraw?: {
+    asymmetry: string
+    basisPoints: string
+    impermanentLossProtection: string
+    liquidityUnits: string
+    memo: string
+    networkFees: Coin[]
+  }
+}
+
+interface Actions {
+  date: string
+  height: string
+  in: InOut[]
+  out: InOut[]
+  metadata?: Metadata
+  pools: string[]
+  status: 'success' | 'pending'
+  type: 'swap' | 'addLiquidity' | 'withdraw' | 'donate' | 'refund' | 'switch'
+}
+
+export interface ActionsResponse {
+  actions: Actions[]
+  count: string
+  meta: {
+    nextPageToken: string
+    prevPageToken: string
+  }
+}
+
+interface ExtraMetadata {
+  method: string
+  liquidity?: Liquidity
+  swap?: Swap
+}
+
+const getLiquidityType = (pool: string): LiquidityType => (pool.includes('/') ? 'Savers' : 'LP')
+
+const getSwapType = (memo: string): SwapType => {
+  const regex = /:\d+\/\d+\/\d+:/
+  return regex.test(memo) ? 'Streaming' : 'Standard'
+}
+
+export interface ParserArgs {
+  midgardUrl: string
+}
+
+export class Parser implements SubParser<Tx> {
+  private readonly axiosMidgard: Axios
+
+  constructor(args: ParserArgs) {
+    this.axiosMidgard = axios.create({ baseURL: args.midgardUrl })
+  }
+
+  async parse(tx: Tx): Promise<TxSpecific | undefined> {
+    const opReturn = tx.vout.find(vout => vout.opReturn)?.opReturn
+    if (!opReturn) return
+
+    const regex = /OP_RETURN \(([^)]+)\)/
```

Comment:
Ditto opReTurnData regex terminology, perhaps outside of this scope

Reply:
Also very nitpicky but can use capture groups if we want to make this cleaner and more declarative, e.g:

```typescript
"OP_RETURN ($+:THOR.RUNE:thor1h7jfggcpqvc4dz546wm9etlgsas0x6su3q490w)".match(/OP_RETURN \((?<memo>[^)]+)\)/).groups.memo
// "$+:THOR.RUNE:thor1h7jfggcpqvc4dz546wm9etlgsas0x6su3q490w"

"foobar".match(/OP_RETURN \((?<memo>[^)]+)\)/)?.groups?.memo

// undefined
```

Reply:
https://github.com/shapeshift/web/pull/5986/commits/efc20f05273df748fdf20442cea2874f23639a30